File: app/src/main/java/org/schabi/newpipe/player/mediaitem/MediaItemTag.java
package org.schabi.newpipe.player.mediaitem;

import android.net.Uri;

import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.MediaItem.RequestMetadata;
import com.google.android.exoplayer2.MediaMetadata;
import com.google.android.exoplayer2.Player;

import org.schabi.newpipe.extractor.stream.AudioStream;
import org.schabi.newpipe.extractor.stream.StreamInfo;
import org.schabi.newpipe.extractor.stream.StreamType;
import org.schabi.newpipe.extractor.stream.VideoStream;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Metadata container and accessor used by player internals.
 *
 * This interface ensures consistency of fetching metadata on each stream,
 * which is encapsulated in a {@link MediaItem} and delivered via ExoPlayer's
 * {@link Player.Listener} on event triggers to the downstream users.
 **/
public interface MediaItemTag {

    List<Exception> getErrors();

    int getServiceId();

    String getTitle();

    String getUploaderName();

    long getDurationSeconds();

    String getStreamUrl();

    String getThumbnailUrl();

    String getUploaderUrl();

    StreamType getStreamType();

    @NonNull
    default Optional<StreamInfo> getMaybeStreamInfo() {
        return Optional.empty();
    }

    @NonNull
    default Optional<Quality> getMaybeQuality() {
        return Optional.empty();
    }

    @NonNull
    default Optional<AudioTrack> getMaybeAudioTrack() {
        return Optional.empty();
    }

    <T> Optional<T> getMaybeExtras(@NonNull Class<T> type);

    <T> MediaItemTag withExtras(@NonNull T extra);

    @NonNull
    static Optional<MediaItemTag> from(@Nullable final MediaItem mediaItem) {
        return Optional.ofNullable(mediaItem)
                .map(item -> item.localConfiguration)
                .map(localConfiguration -> localConfiguration.tag)
                .filter(MediaItemTag.class::isInstance)
                .map(MediaItemTag.class::cast);
    }

    @NonNull
    default String makeMediaId() {
        return UUID.randomUUID().toString() + "[" + getTitle() + "]";
    }

    @NonNull
    default MediaItem asMediaItem() {
        final String thumbnailUrl = getThumbnailUrl();
        final MediaMetadata mediaMetadata = new MediaMetadata.Builder()
                .setArtworkUri(thumbnailUrl == null ? null : Uri.parse(thumbnailUrl))
                .setArtist(getUploaderName())
                .setDescription(getTitle())
                .setDisplayTitle(getTitle())
                .setTitle(getTitle())
                .build();

        final RequestMetadata requestMetaData = new RequestMetadata.Builder()
                .setMediaUri(Uri.parse(getStreamUrl()))
                .build();

        return MediaItem.fromUri(getStreamUrl())
                .buildUpon()
                .setMediaId(makeMediaId())
                .setMediaMetadata(mediaMetadata)
                .setRequestMetadata(requestMetaData)
                .setTag(this)
                .build();
    }

    final class Quality {
        @NonNull
        private final List<VideoStream> sortedVideoStreams;

        /** Invariant: Index exists in sortedVideoStreams. */
        private final int selectedVideoStreamIndex;


        /** Create a new video Quality. The index must be valid in `sortedVideoStreams`.
         *
         * @param sortedVideoStreams
         * @param selectedVideoStreamIndex
         * @throws ArrayIndexOutOfBoundsException if index does not exist in `sortedVideoStreams`
         */
        public Quality(@NonNull final List<VideoStream> sortedVideoStreams,
                        final int selectedVideoStreamIndex) {
            if  (selectedVideoStreamIndex < 0
                    || selectedVideoStreamIndex >= sortedVideoStreams.size()) {
                throw new ArrayIndexOutOfBoundsException(
                        "selectedVideoStreamIndex does not exist in sortedVideoStreams");
            }
            this.sortedVideoStreams = sortedVideoStreams;
            this.selectedVideoStreamIndex = selectedVideoStreamIndex;
        }

        @NonNull
        public List<VideoStream> getSortedVideoStreams() {
            return sortedVideoStreams;
        }

        public int getSelectedVideoStreamIndex() {
            return selectedVideoStreamIndex;
        }

        @NonNull
        public VideoStream getSelectedVideoStream() {
            return sortedVideoStreams.get(selectedVideoStreamIndex);
        }
    }

    final class AudioTrack {
        @NonNull
        private final List<AudioStream> audioStreams;
        /** Invariant: Index exists in audioStreams. */
        private final int selectedAudioStreamIndex;

        /** Create a new AudioTrack. The index must be valid in `audioStreams`.
         *
         * @param audioStreams
         * @param selectedAudioStreamIndex
         * @throws ArrayIndexOutOfBoundsException if index does not exist in audioStreams.
         */
        public AudioTrack(@NonNull final List<AudioStream> audioStreams,
                           final int selectedAudioStreamIndex) {
            if  (selectedAudioStreamIndex < 0
                    || selectedAudioStreamIndex >= audioStreams.size()) {
                throw new ArrayIndexOutOfBoundsException(
                        "selectedAudioStreamIndex does not exist in audioStreams");
            }
            this.audioStreams = audioStreams;
            this.selectedAudioStreamIndex = selectedAudioStreamIndex;
        }

        @NonNull
        public List<AudioStream> getAudioStreams() {
            return audioStreams;
        }

        public int getSelectedAudioStreamIndex() {
            return selectedAudioStreamIndex;
        }

        @NonNull
        public AudioStream getSelectedAudioStream() {
            return audioStreams.get(selectedAudioStreamIndex);
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/player/mediaitem/MediaItemTag.kt
@file:OptIn(androidx.media3.common.util.UnstableApi::class)

package org.schabi.newpipe.player.mediaitem

import android.net.Uri
import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import androidx.media3.common.Player
import java.util.UUID
import org.schabi.newpipe.extractor.stream.AudioStream
import org.schabi.newpipe.extractor.stream.StreamInfo
import org.schabi.newpipe.extractor.stream.StreamType
import org.schabi.newpipe.extractor.stream.VideoStream

/**
 * Metadata container and accessor used by player internals.
 *
 * This interface ensures consistency of fetching metadata on each stream,
 * which is encapsulated in a [MediaItem] and delivered via ExoPlayer's
 * [Player.Listener] on event triggers to the downstream users.
 */
interface MediaItemTag {
    val errors: List<Exception>
    val serviceId: Int
    val title: String?
    val uploaderName: String?
    val durationSeconds: Long
    val streamUrl: String
    val thumbnailUrl: String?
    val uploaderUrl: String?
    val streamType: StreamType?

    val maybeStreamInfo: StreamInfo? get() = null
    val maybeQuality: Quality? get() = null
    val maybeAudioTrack: AudioTrack? get() = null

    fun <T : Any> getMaybeExtras(type: Class<T>): T?
    fun <T : Any> withExtras(extra: T): MediaItemTag

    fun makeMediaId(): String {
        return UUID.randomUUID().toString() + "[" + title + "]"
    }

    fun asMediaItem(): MediaItem {
        val thumbnailUrl = thumbnailUrl
        val mediaMetadata = MediaMetadata.Builder()
            .setArtworkUri(if (thumbnailUrl == null) null else Uri.parse(thumbnailUrl))
            .setArtist(uploaderName)
            .setDescription(title)
            .setDisplayTitle(title)
            .setTitle(title)
            .build()

        val requestMetaData = MediaItem.RequestMetadata.Builder()
            .setMediaUri(Uri.parse(streamUrl))
            .build()

        return MediaItem.fromUri(streamUrl)
            .buildUpon()
            .setMediaId(makeMediaId())
            .setMediaMetadata(mediaMetadata)
            .setRequestMetadata(requestMetaData)
            .setTag(this)
            .build()
    }

    class Quality private constructor(
        val sortedVideoStreams: List<VideoStream>,
        val selectedVideoStreamIndex: Int
    ) {
        val selectedVideoStream: VideoStream?
            get() = if (selectedVideoStreamIndex < 0 || selectedVideoStreamIndex >= sortedVideoStreams.size) {
                null
            } else {
                sortedVideoStreams[selectedVideoStreamIndex]
            }

        companion object {
            @JvmStatic
            fun of(sortedVideoStreams: List<VideoStream>, selectedVideoStreamIndex: Int): Quality {
                return Quality(sortedVideoStreams, selectedVideoStreamIndex)
            }
        }
    }

    class AudioTrack private constructor(
        val audioStreams: List<AudioStream>,
        val selectedAudioStreamIndex: Int
    ) {
        val selectedAudioStream: AudioStream?
            get() = if (selectedAudioStreamIndex < 0 || selectedAudioStreamIndex >= audioStreams.size) {
                null
            } else {
                audioStreams[selectedAudioStreamIndex]
            }

        companion object {
            @JvmStatic
            fun of(audioStreams: List<AudioStream>, selectedAudioStreamIndex: Int): AudioTrack {
                return AudioTrack(audioStreams, selectedAudioStreamIndex)
            }
        }
    }

    companion object {
        @JvmStatic
        fun from(mediaItem: MediaItem?): MediaItemTag? {
            return mediaItem?.localConfiguration?.tag as? MediaItemTag
        }
    }
}
